### PR TITLE
Allow replacing subcomponents

### LINKED ISSUE #466: Allow replacing `@ContributesSubcomponent`
I didn't implement this feature, because it seemed unnecessary. But there's actually at least one reasonable use case: contributing a new subcomponent with different scope annotations, e.g.
```kotlin
@ContributesSubcompnent(scope = Unit::class, parentScope = Any::class)
interface Subcomponent

@ContributesSubcompnent(scope = Unit::class, parentScope = Any::class, replaces = [Subcomponent::class])
@Singleton
interface OtherSubcomponent
```

### PR DESCRIPTION
**Pull request by vRallev**
Fixes #466

### CODE REVIEW

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt [hunk 1]
```diff
@@ -101,9 +106,51 @@ internal class ContributesSubcomponentHandlerGenerator(
     module: ModuleDescriptor,
     projectFiles: Collection<KtFile>
   ): Collection<GeneratedFile> {
+    if (isFirstRound) {
```
Comment (vRallev) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt:109]: It was unnecessary to do this work over and over again.

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt [hunk 2]
```diff
@@ -1628,6 +1628,223 @@ class ContributesSubcomponentHandlerGeneratorTest {
...
+          .toList()
+      ).containsExactly(subcomponentInterface2)
+    }
+  }
+
+  @Test fun `a contributed subcomponent can be replaced by a generated subcomponent`() {
+    // This test verifies an edge case where we generate the code for a contributed subcomponent,
+    // but then later in a new code generation round another code generator generates code that
+    // is supposed to replace the already generated subcomponent. We can't revert the code and
+    // don't want to support that use case.
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt:1780]: Kinda proud that I was thinking of this 😅

**Comment by vRallev** [comment]
Tested it internally and it works 🎉

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt [hunk 3]
```diff
@@ -101,9 +105,51 @@ internal class ContributesSubcomponentHandlerGenerator(
...
```
Comment (JoelWilcox) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt:109]: Nit: This `generateCode` method seems to have gotten quite large and it makes what's happening a bit harder to follow. It would be helpful to break some of its logic out to new helper methods. e.g.
```
if (isFirstRound) {
  isFirstRound = false
  populateInitialContributions(module)
  populateReplacedContributions(module)
}
```

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentGeneratorTest.kt [hunk 4]
```diff
@@ -463,6 +464,72 @@ class ContributesSubcomponentGeneratorTest {
...
+        replaces = [SubcomponentInterface1::class]
+      )
+      interface SubcomponentInterface2
+      """
+    ) {
+      assertThat(subcomponentInterface1.hintSubcomponent?.java).isEqualTo(subcomponentInterface1)
+      assertThat(subcomponentInterface1.hintSubcomponentParentScope).isEqualTo(Unit::class)
+
+      assertThat(subcomponentInterface2.hintSubcomponent?.java).isEqualTo(subcomponentInterface2)
+      assertThat(subcomponentInterface2.hintSubcomponentParentScope).isEqualTo(Int::class)
```
Comment (JoelWilcox) [inline compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentGeneratorTest.kt:492]: At first glance, this test makes it look like both subcomponents are contributed successfully, rather than the first being replaced. I was expecting to see some sort of `assertFalse` style assert confirming that `SubcomponentInterface1` was not contributed. Is the reasoning that these second two asserts would fail if `SubcomponentInterface1` hadn't been replaced?

edit: I see that the test I was expecting is included in ContributesSubcomponentHandlerGeneratorTest.kt. It might be helpful to name this one as something like `a contributed subcomponent can be configured to replace another subcomponent`, if the intention is to make sure including `replaces = []` doesn't break anything.

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt [hunk 5]
```diff
@@ -1628,6 +1628,151 @@ class ContributesSubcomponentHandlerGeneratorTest {
...
+        interface SubcomponentInterface2
+        
+        @MergeComponent(Unit::class)
+        interface ComponentInterface
+      """
+    ) {
+      val annotation = subcomponentInterface2.anvilComponent(componentInterface)
+        .getAnnotation(MergeSubcomponent::class.java)
+
+      assertThat(annotation.exclude).isEmpty()
```
Comment (JoelWilcox) [inline compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt:1715]: I could use a little clarification on this test as well. The test name makes me think that `SubcomponentInterface1` cannot exclude Dagger modules, component interfaces and bindings (and doing so would result in a failure), but the assert appears to be making sure that `SubcomponentInterface2` doesn't exclude anything. Is this basically saying that all excluded items from the replaced subcomponent will be lost?

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt [hunk 6]
```diff
@@ -1773,6 +1773,78 @@ class ContributesSubcomponentHandlerGeneratorTest {
     }
   }
 
+  @Test fun `a contributed subcomponent can be replaced by a generated subcomponent`() {
```
Comment (JoelWilcox) [inline compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt:1776]: This name sounds more like the use-case is supported to me. What about `a previously generated contributed subcomponent cannot be replaced in a later round of generations`? Or at least `can` -> `can't`

**Review by JoelWilcox — COMMENTED** [review]
Overall lgtm! Mostly interested in some test clarifications

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt [hunk 7]
```diff
@@ -90,6 +91,9 @@ internal class ContributesSubcomponentHandlerGenerator(
 
   private val triggers = mutableListOf<Trigger>()
   private val contributions = mutableListOf<Contribution>()
+  private val replacedContributions = mutableSetOf<ClassReference>()
```
[the remark this answers is not in the document]
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt:94]: Agree, I changed it to `replacedReferences`. That's good enough.

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt [hunk 8]
```diff
@@ -101,9 +105,51 @@ internal class ContributesSubcomponentHandlerGenerator(
...
+      isFirstRound = false
+
+      // Find all contributed subcomponents from precompiled dependencies and generate the
+      // necessary code eventually if there's a trigger.
+      contributions += classScanner
+        .findContributedClasses(
+          module = module,
+          packageName = HINT_SUBCOMPONENTS_PACKAGE_PREFIX,
+          annotation = contributesSubcomponentFqName,
+          // Don't use trigger.scope, because it refers to the parent scope of the
```
[the remark this answers is not in the document]
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt:119]: Good catch

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt (hunk 3 above)
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt:109]: Good idea, done.

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentGeneratorTest.kt (hunk 4 above)
Comment (vRallev) [reply compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentGeneratorTest.kt:492]: Very good suggestion!

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt (hunk 5 above)
Comment (vRallev) [reply compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt:1715]: I changed the method name to `a replaced subcomponent's exclusions for Dagger modules, component interfaces and bindings are ignored`.

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt (hunk 6 above)
Comment (vRallev) [reply compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt:1776]: I forgot to update the name. I did TDD and initially planned to implement the functionality, but it was quite tricky and I stopped. Since this is such an edge case I added an explicit check to ensure it isn't supported and then forgot to update the method name.

**Review by JoelWilcox — APPROVED** [review]
Looks great!
